### PR TITLE
Supported some data for shulker boxes

### PR DESCRIPTION
**Pull request by liach**
List of things I did: 
1. Supported inventory lens for shulker boxes. (I need these function)
2. Moved `setCustomDisplayName` to `MixinTileEntityLockableLoot`, the superclass of mixins for chests, hoppers, and shulker boxes.
3. Supported colored data for shulker box block entites (although it is not possible for items, but it is possible for block entities by the way vanilla stores data)

This is my first pull on SpongeCommon, so may be erroneous since this is my first time working with data processors and mixin.

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/processor/data/DyeableDataProcessor.java [hunk 1]
```diff
@@ -95,6 +98,9 @@ public boolean supports(DataHolder dataHolder) {
                 } else {
                     return Optional.of(new SpongeDyeableData((DyeColor) (Object) EnumDyeColor.byMetadata(((ItemStack) dataHolder).getItemDamage())));
                 }
+            } else if (dataHolder instanceof TileEntityShulkerBox) {
+                return Optional.ofNullable(((IMixinTileEntityShulkerBox) dataHolder).getColor()).map(enumType -> (DyeColor) (Object) enumType).map
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/DyeableDataProcessor.java]: Make this separately lined for legibility.

File: src/main/java/org/spongepowered/common/data/processor/data/DyeableDataProcessor.java [hunk 2]
```diff
@@ -79,7 +82,7 @@ protected DyeableData createManipulator() {
     @Override
     public boolean supports(DataHolder dataHolder) {
         return dataHolder instanceof EntityWolf || dataHolder instanceof EntitySheep || dataHolder instanceof ItemStack &&
-                isDyeable(((ItemStack) dataHolder).getItem());
+                isDyeable(((ItemStack) dataHolder).getItem()) || dataHolder instanceof TileEntityShulkerBox;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/DyeableDataProcessor.java]: Could really just make a separate data processor instead and register that one.

**Comment by liach** [comment]
@kashike Ping the great reviewer to give me some idea.

File: src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityLockableLoot.java [hunk 3]
```diff
@@ -45,4 +46,9 @@ public DataContainer toContainer() {
         return container;
     }
 
+    @Override
+    public void setCustomDisplayName(String name) {
+        ((TileEntityLockableLoot) (Object) this).setCustomName(name);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityLockableLoot.java]: Shadow this method instead.

File: src/main/java/org/spongepowered/common/data/processor/value/tileentity/ShulkerBoxDyeColorValueProcessor.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+public class ShulkerBoxDyeColorValueProcessor extends AbstractSpongeValueProcessor<TileEntityShulkerBox, DyeColor, Value<DyeColor>> {
+
+    public ShulkerBoxDyeColorValueProcessor() {
+        super(TileEntityShulkerBox.class, Keys.DYE_COLOR);
+    }
+
+    @Override
+    protected Value<DyeColor> constructValue(DyeColor actualValue) {
+        return SpongeValueFactory.getInstance().createValue(Keys.DYE_COLOR, actualValue, DyeColors.PURPLE);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/data/processor/value/tileentity/ShulkerBoxDyeColorValueProcessor.java]: `new SpongeValue<>(this.key, DyeColors.PURPLE, actualValue);`

File: src/main/java/org/spongepowered/common/data/processor/value/tileentity/ShulkerBoxDyeColorValueProcessor.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Override
+    protected Optional<DyeColor> getVal(TileEntityShulkerBox container) {
+        return Optional.ofNullable(((IMixinTileEntityShulkerBox) container).getColor())
+                .map(enumValue -> (DyeColor) (Object) enumValue);
+    }
+
+    @Override
+    protected ImmutableValue<DyeColor> constructImmutableValue(DyeColor value) {
+        return constructValue(value).asImmutable();
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/data/processor/value/tileentity/ShulkerBoxDyeColorValueProcessor.java]: `new ImmutableSpongeValue<>(this.key, DyeColors.PURPLE, value)`

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityLockableLoot.java (hunk 3 above)
Comment (liach) [reply src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityLockableLoot.java]: Good point. The old code used casts instead of shadowing and I copied it.

File: src/main/java/org/spongepowered/common/data/processor/value/tileentity/ShulkerBoxDyeColorValueProcessor.java (hunk 4 above)
Comment (liach) [reply src/main/java/org/spongepowered/common/data/processor/value/tileentity/ShulkerBoxDyeColorValueProcessor.java]: Will fix. Copied from ItemStackDyeColorValueProcessor.

**Comment by liach** [comment]
@kashike Is my accessor mixin right? My Minecraft dev plugin report that "I cannot reference a mixin class directly".

**Review by gabizou — CHANGES_REQUESTED** [review]
Some minor things.

File: src/main/java/org/spongepowered/common/data/processor/data/DyeableDataProcessor.java (hunk 2 above)
Comment (liach) [reply src/main/java/org/spongepowered/common/data/processor/data/DyeableDataProcessor.java]: Can it really support more than one data processor?

Comment (gabizou) [reply src/main/java/org/spongepowered/common/data/processor/data/DyeableDataProcessor.java]: Yes... That's the whole point of the data processors. If anything, multiple processors focusing on doing "one thing well" is part of what the processors are written for, and why there's multiple value processors.

**Comment by liach** [comment]
@gabizou @kashike Fixed.

**Comment by liach** [comment]
For my plugin's sake, can people take a look at this?

File: src/main/java/org/spongepowered/common/data/processor/data/DyeableDataProcessor.java [hunk 6]
```diff
@@ -65,7 +65,7 @@ public static boolean isDyeable(Item item) {
 
         Block block = Block.getBlockFromItem(item);
         return block != null
-            && (block.equals(Blocks.WOOL)
+                && (block.equals(Blocks.WOOL)
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/DyeableDataProcessor.java]: These are now unnecessary changes and are considered commit noise.

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.data.processor.data.tileentity;
+
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java:26]: Extra line.

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    @Override
+    public boolean supports(DataHolder dataHolder) {
+        return dataHolder instanceof TileEntityShulkerBox;
+    }
+
+    @Override
+    public Optional<DyeableData> from(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            return Optional.ofNullable(((IMixinTileEntityShulkerBox) dataHolder).getColor())
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java]: Yeah.... this is functionally ok, but it's very wasteful. Just use traditional `if (color != null)` logic. It reads better.

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+                    .map(SpongeDyeableData::new);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<DyeableData> fill(DataContainer container, DyeableData dyeableData) {
+        Optional<String> id = container.getString(Keys.DYE_COLOR.getQuery());
+        if (id.isPresent()) {
+            dyeableData.set(Keys.DYE_COLOR, Sponge.getGame().getRegistry().getType(DyeColor.class, id.get()).get());
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java]: Multiple points of failure in one line. Split this up into multiple lines to be able to throw exceptions as necessary.

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+    @Override
+    public DataTransactionResult set(DataHolder dataHolder, DyeableData manipulator, MergeFunction function) {
+        if (!supports(dataHolder)) {
+            return DataTransactionResult.failResult(manipulator.type().asImmutable());
+        }
+        DyeableData original = from(dataHolder).get();
+        DyeableData merged = function.merge(original, manipulator);
+        ((IMixinTileEntityShulkerBox) dataHolder).setColor((EnumDyeColor) (Object) merged.type().get());
+        return DataTransactionResult.successReplaceResult(original.type().asImmutable(), merged.type().asImmutable());
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java]: While this method does read ok, it is wasting a lot of resources and generating unnecessary objects with each line.

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        if (id.isPresent()) {
+            dyeableData.set(Keys.DYE_COLOR, Sponge.getGame().getRegistry().getType(DyeColor.class, id.get()).get());
+            return Optional.of(dyeableData);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public DataTransactionResult set(DataHolder dataHolder, DyeableData manipulator, MergeFunction function) {
+        if (!supports(dataHolder)) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java]: This is already checked for you. Unnecessary.

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public DataTransactionResult remove(DataHolder dataHolder) {
+        IMixinTileEntityShulkerBox box = (IMixinTileEntityShulkerBox) dataHolder;
+        DyeColor color = (DyeColor) (Object) box.getColor();
+        box.setColor(null);
+        return DataTransactionResult.successRemove(new ImmutableSpongeValue<>(Keys.DYE_COLOR, DyeColors.PURPLE, color));
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java]: The value could be cached through the ImmutableDataCachingUtil

File: src/main/java/org/spongepowered/common/data/processor/value/tileentity/ShulkerBoxDyeColorValueProcessor.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Override
+    protected boolean set(TileEntityShulkerBox container, DyeColor value) {
+        ((IMixinTileEntityShulkerBox) container).setColor((EnumDyeColor) (Object) value);
+        return true;
+    }
+
+    @Override
+    protected Optional<DyeColor> getVal(TileEntityShulkerBox container) {
+        return Optional.ofNullable(((IMixinTileEntityShulkerBox) container).getColor())
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/tileentity/ShulkerBoxDyeColorValueProcessor.java]: Same as before. Unnecessary extra operations for minimal gain.

File: src/main/java/org/spongepowered/common/data/processor/value/tileentity/ShulkerBoxDyeColorValueProcessor.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+                .map(enumValue -> (DyeColor) (Object) enumValue);
+    }
+
+    @Override
+    protected ImmutableValue<DyeColor> constructImmutableValue(DyeColor value) {
+        return new ImmutableSpongeValue<>(Keys.DYE_COLOR, DyeColors.PURPLE, value);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/tileentity/ShulkerBoxDyeColorValueProcessor.java]: Can be cached.

File: src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityChest.java [hunk 15]
```diff
@@ -98,7 +97,9 @@ public void onUpdate(CallbackInfo ci) {
         }
     }
 
-    @Inject(method = "openInventory", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/World;addBlockEvent(Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/block/Block;II)V"), cancellable = true)
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityChest.java]: What is this actually changing? Nothing in this PR looks to really touch chests...

File: src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityHopper.java [hunk 16]
```diff
@@ -78,7 +77,8 @@ public void onConstructed(CallbackInfo ci) {
         this.lens = new GridInventoryLensImpl(0, 5, 1, 5, this.slots);
     }
 
-    @Inject(method = "putDropInInventoryAllSlots", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/item/EntityItem;getEntityItem()Lnet/minecraft/item/ItemStack;"))
+    @Inject(method = "putDropInInventoryAllSlots",
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityHopper.java]: change line noise.

File: src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityShulkerBox.java [hunk 17]
```diff
@@ -24,13 +24,77 @@
...
+
+import java.util.List;
+import java.util.Optional;
+
+import javax.annotation.Nullable;
 
 @NonnullByDefault
 @Mixin(TileEntityShulkerBox.class)
+@Implements(@Interface(iface = MinecraftInventoryAdapter.class, prefix = "inventory$"))
 public abstract class MixinTileEntityShulkerBox extends MixinTileEntityLockableLoot implements ShulkerBox {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityShulkerBox.java]: For the sake of it, you can also just add the implementing of `IMixinTileEntityShulkerBox`.

**Review by gabizou — CHANGES_REQUESTED** [review]
Issues with implementation.

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java (hunk 10 above)
Comment (liach) [reply src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java]: Copied from `DyeableDataProcessor`. I hope you give me an example of good data processor and I will imitate the good one.

Comment (gabizou) [reply src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java]: Don't look at an example, think about what objects are being created, what processes those involve, etc. How many duplicate objects or duplicate processes are being called. Yes, you have a merged DyeableData, but is it necessary to constantly call type() twice when calling it once would suffice? Or calling it once to just get the immutable variant from cache?

Comment (liach) [reply src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java]: So `successNoResult`?

File: src/main/java/org/spongepowered/common/data/processor/value/tileentity/ShulkerBoxDyeColorValueProcessor.java (hunk 14 above)
Comment (liach) [reply src/main/java/org/spongepowered/common/data/processor/value/tileentity/ShulkerBoxDyeColorValueProcessor.java]: Use a `Map` or `ImmutableDataCachingUtil`?

Comment (liach) [reply src/main/java/org/spongepowered/common/data/processor/value/tileentity/ShulkerBoxDyeColorValueProcessor.java]: Oh now I knows

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+                    .map(SpongeDyeableData::new);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<DyeableData> fill(DataContainer container, DyeableData dyeableData) {
+        Optional<String> id = container.getString(Keys.DYE_COLOR.getQuery());
+        if (id.isPresent()) {
+            dyeableData.set(Keys.DYE_COLOR, Sponge.getGame().getRegistry()
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java]: When I said to split this line up, I meant to assign bits to local variables, throw exceptions on the optionals (using `orElseThrow`) etc.

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+            return Optional.of(dyeableData);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public DataTransactionResult set(DataHolder dataHolder, DyeableData manipulator, MergeFunction function) {
+        DyeableData original = from(dataHolder).get();
+        DyeableData merged = function.merge(original, manipulator);
+        ((IMixinTileEntityShulkerBox) dataHolder).setColor((EnumDyeColor) (Object) merged.type().get());
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java]: You can replace the `merged.type()` as a local variable to use in two places.

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+        return DataTransactionResult.successReplaceResult(original.type().asImmutable(), merged.type().asImmutable());
+    }
+
+    @Override
+    public Optional<ImmutableDyeableData> with(Key<? extends BaseValue<?>> key, Object value, ImmutableDyeableData immutable) {
+        if (key.equals(Keys.DYE_COLOR)) {
+            return Optional.of(new ImmutableSpongeDyeableData((DyeColor) value));
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java]: Can't this be cached?

**Review by gabizou — CHANGES_REQUESTED** [review]
Caching util can be your friend. Maybe reduce single line multiple operations with local variables to make it clear what's being used where and when and how many times in some areas.

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java (hunk 20 above)
Comment (liach) [reply src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java]: Change is made, pushing... The gradle task costs time a lot.

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java (hunk 7 above)
Comment (liach) [reply src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java:26]: Fixed. 😄

**Comment by liach** [comment]
I announce that I've done your requests for now. :smiley:

**Comment by liach** [comment]
This is also inventory. @kashike

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    @Override
+    public boolean supports(DataHolder dataHolder) {
+        return dataHolder instanceof TileEntityShulkerBox;
+    }
+
+    @Override
+    public Optional<DyeableData> from(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java]: This also isn't necessary since supports is checked by the delegate.

File: src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityShulkerBox.java [hunk 22]
```diff
@@ -24,13 +24,78 @@
...
+    public void onConstructed(CallbackInfo ci) {
+        this.fabric = new DefaultInventoryFabric(this);
+        this.slots = new SlotCollection.Builder().add(27).build();
+        this.lens = new GridInventoryLensImpl(0, 9, 3, 9, this.slots);
+    }
+
+    @Override
+    public void supplyVanillaManipulators(List<DataManipulator<?, ?>> manipulators) {
+        super.supplyVanillaManipulators(manipulators);
+        get(DyeableData.class)
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/tileentity/MixinTileEntityShulkerBox.java]: Use faster methods, should not use data processors when possible.

**Review by gabizou — CHANGES_REQUESTED** [review]
More minor changes for performance.

**Comment by liach** [comment]
@gabizou Fixed performance issues.

**Comment by gabizou** [comment]
@Faithcaio or @Mumfrey can you review the inventory changes with this PR?

Otherwise, the data bits look good.

**Comment by liach** [comment]
This pull simply added lens support for shulker boxes, and moved some stuff to the common superclass of chest, hopper, and shulker boxes

**Comment by liach** [comment]
Any update after one week?

**Comment by liach** [comment]
Any update after 2 weeks?

**Comment by Faithcaio** [comment]
looks fine on the inventory side of things

**Comment by liach** [comment]
Any conclusion?

**Comment by gabizou** [comment]
Can you squash your commits into 1 comment?

**Comment by liach** [comment]
Do you mean to squash all these into 1 commit?

**Comment by gabizou** [comment]
Yes.

**Comment by liach** [comment]
Done.

**Comment by liach** [comment]
:trollface:

**Comment by liach** [comment]
@gabizou Any updates?

**Comment by Faithcaio** [comment]
@liach I included some of your changes in: https://github.com/SpongePowered/SpongeCommon/commit/ba2a27ffa83ee972a024407921fbd36c01e768c0

So you're going to have a few conflicts

**Comment by gabizou** [comment]
Likewise, I still have an abstract method error now in development environment when I tried to right click a shulker box.

**Comment by Faithcaio** [comment]
@gabizou the same getRootLens() one? It was working for me.

**Comment by parlough** [comment]
@liach Can you update this?

File: src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+import org.spongepowered.api.registry.RegistryException;
+import org.spongepowered.common.data.ImmutableDataCachingUtil;
+import org.spongepowered.common.data.manipulator.mutable.SpongeDyeableData;
+import org.spongepowered.common.data.processor.common.AbstractSingleDataProcessor;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+import org.spongepowered.common.interfaces.block.tile.IMixinTileEntityShulkerBox;
+
+import java.util.Optional;
+
+public class ShulkerBoxDyeableDataProcessor extends AbstractSingleDataProcessor<DyeColor, Value<DyeColor>, DyeableData, ImmutableDyeableData> {
```
Comment (JBYoshi) [inline src/main/java/org/spongepowered/common/data/processor/data/tileentity/ShulkerBoxDyeableDataProcessor.java:51]: Use `AbstractTileEntitySingleDataProcessor`. It will let you delete the `ValueProcessor`; just remove the `registerValueProcessor()` entry in `DataRegistrar` and change the `registerDataProcessorAndImpl()` call with `registerDualProcessor()` (leave it in the same location).

**Review by JBYoshi — CHANGES_REQUESTED** [review]

**Comment by liach** [comment]
After reading the conflicts, I think I'd better remake a pull. I will leave this as a reference for now.

**Comment by parlough** [comment]
@liach I am not sure what is still needed for the PR, but since you said you will remake a PR if you come back to this anyway I am going to go ahead and close it for now. If you wouldn't mind figuring out what is still needed to be done, please respond here(or create a SC issue), as then you or someone else(me, faith, etc) could finish it when we get the chance. Thanks.

I am tracking this in a project so I will come back to it soon.

**Comment by liach** [comment]
@Meronat Sorry, I should have closed this pull thousand years earlier.
